[pytorch][PR] introduce WeakHashRef

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115216

We would like weak dictionaries that have `torch.ScriptObject` keys. Similar to tensors, we need to override the behavior of the ref to dot he right thing under comparison.

This change also makes it so that WeakIdKeyDictionary works with a pluggable ref_type.

Differential Revision: [D51828205](https://our.internmc.facebook.com/intern/diff/D51828205/)

suo committed:
[pytorch][PR] introduce WeakscriptObjectRef

We would like weak dictionaries that have `torch.ScriptObject` keys. Similar to tensors, we need to override the behavior of the ref to dot he right thing under comparison.

This change also makes it so that WeakIdKeyDictionary works with a pluggable ref_type.

Differential Revision: [D51828205](https://our.internmc.facebook.com/intern/diff/D51828205/)

[ghstack-poisoned]

suo committed:
Update on "[pytorch][PR] introduce WeakscriptObjectRef"


We would like weak dictionaries that have `torch.ScriptObject` keys. Similar to tensors, we need to override the behavior of the ref to dot he right thing under comparison.

This change also makes it so that WeakIdKeyDictionary works with a pluggable ref_type.

Differential Revision: [D51828205](https://our.internmc.facebook.com/intern/diff/D51828205/)

[ghstack-poisoned]

suo committed:
Update on "[pytorch][PR] introduce WeakscriptObjectRef"


We would like weak dictionaries that have `torch.ScriptObject` keys. Similar to tensors, we need to override the behavior of the ref to dot he right thing under comparison.

This change also makes it so that WeakIdKeyDictionary works with a pluggable ref_type.

Differential Revision: [D51828205](https://our.internmc.facebook.com/intern/diff/D51828205/)

[ghstack-poisoned]

suo commented:
ignore the lint failure; https://github.com/pytorch/pytorch/pull/115233 should move that check to use a different thing

albanD commented on test/test_weak.py:
```diff
@@ -562,5 +570,340 @@ def test_pop(self):
+            torch.ops.load_library(str(lib_file_path))
+
+        unittest.TestCase.__init__(self, *args, **kw)
+        self.reference = self._reference().copy()
```
This will be shared by all tests. Shouldn't this be in setUp() ?

suo replied:
this is part of the code copied from cpython; I'd prefer not to change unless necessary lol

albanD commented on test/test_weak.py:
```diff
@@ -562,5 +570,340 @@ def test_pop(self):
+    def test_read(self):
+        # Test for read only operations on mapping
+        p = self._empty_mapping()
+        p1 = dict(p)  # workaround for singleton objects
```
What does this mean?

albanD reviewed (commented):
Tests are a bit confusing haha

suo commented on docs/source/conf.py:
```diff
@@ -3487,6 +3487,7 @@
     "WeakIdKeyDictionary",
     "WeakIdRef",
     "WeakTensorKeyDictionary",
+    "WeakScriptObjectRef",
```
I'll make WeakScriptObjectRef hidden per discussion

suo commented on torch/utils/weak.py:
```diff
@@ -80,13 +80,42 @@ def __eq__(self, other):
+# same IValue as the equal.
+class WeakScriptObjectRef(weakref.ref):
+    __slots__ = ['_id']
+
```
oh I see, you wanted to make it more generic

suo committed:
Update on "[pytorch][PR] introduce WeakscriptObjectRef"


We would like weak dictionaries that have `torch.ScriptObject` keys. Similar to tensors, we need to override the behavior of the ref to dot he right thing under comparison.

This change also makes it so that WeakIdKeyDictionary works with a pluggable ref_type.

Differential Revision: [D51828205](https://our.internmc.facebook.com/intern/diff/D51828205/)

[ghstack-poisoned]

albanD commented on torch/utils/weak.py:
```diff
@@ -4,14 +4,14 @@
 
 
-__all__ = ['TensorWeakRef', 'WeakIdRef', 'WeakIdKeyDictionary', 'WeakTensorKeyDictionary']
+__all__ = ['TensorWeakRef', 'WeakIdRef', 'WeakIdKeyDictionary', 'WeakTensorKeyDictionary', '_WeakHashRef']
```
Don't put it here, this should contain only the public APIs.

albanD reviewed (approved):
Small details on the public API but good to go once the CI is green!

suo committed:
Update on "[pytorch][PR] introduce WeakHashRef"


We would like weak dictionaries that have `torch.ScriptObject` keys. Similar to tensors, we need to override the behavior of the ref to dot he right thing under comparison.

This change also makes it so that WeakIdKeyDictionary works with a pluggable ref_type.

Differential Revision: [D51828205](https://our.internmc.facebook.com/intern/diff/D51828205/)

[ghstack-poisoned]

suo committed:
Update on "[pytorch][PR] introduce WeakHashRef"


We would like weak dictionaries that have `torch.ScriptObject` keys. Similar to tensors, we need to override the behavior of the ref to dot he right thing under comparison.

This change also makes it so that WeakIdKeyDictionary works with a pluggable ref_type.

Differential Revision: [D51828205](https://our.internmc.facebook.com/intern/diff/D51828205/)

[ghstack-poisoned]

suo committed:
Update on "[pytorch][PR] introduce WeakHashRef"


We would like weak dictionaries that have `torch.ScriptObject` keys. Similar to tensors, we need to override the behavior of the ref to dot he right thing under comparison.

This change also makes it so that WeakIdKeyDictionary works with a pluggable ref_type.

Differential Revision: [D51828205](https://our.internmc.facebook.com/intern/diff/D51828205/)

[ghstack-poisoned]
